src/Network/Transport/TCP.hs

```diff
@@ -1717,7 +1743,9 @@ runScheduledAction (ourEndPoint, theirEndPoint) mvar = do
     handleIOException ex vst = do
       -- Release probing resources if probing.
       forM_ (remoteProbing vst) id
-      tryCloseSocket (remoteSocket vst)
```

Comment:
If I understand correctly, reading from the socket can still yield some bytes after shutting down, which could cause a message to be delivered after the `EventConnectionLost`. This was also possible before.

The precise operation we want is not so much shutting down, but terminating the thread which reads the socket. Perhaps, this could be accomplished with an async exception and then waiting on the socket to be closed.